src/Microsoft.Health.Dicom.Core/Features/Update/UpdateInstanceService.cs

```diff
@@ -0,0 +1,202 @@
+{
+    private readonly IFileStore _fileStore;
+    private readonly IMetadataStore _metadataStore;
+    private readonly ILogger<UpdateInstanceService> _logger;
+    private readonly RecyclableMemoryStreamManager _recyclableMemoryStreamManager;
+    private const int LargeObjectsizeInBytes = 1024 * 1024;
+    private const int StageBlockSizeInBytes = 4 * 1024 * 1024;
+
+    public UpdateInstanceService(
+        IFileStore fileStore,
+        IMetadataStore metadataStore,
+        RecyclableMemoryStreamManager recyclableMemoryStreamManager,
+        ILogger<UpdateInstanceService> logger)
+    {
+        _fileStore = EnsureArg.IsNotNull(fileStore, nameof(fileStore));
+        _metadataStore = EnsureArg.IsNotNull(metadataStore, nameof(metadataStore));
+        _logger = EnsureArg.IsNotNull(logger, nameof(logger));
+        _recyclableMemoryStreamManager = EnsureArg.IsNotNull(recyclableMemoryStreamManager, nameof(recyclableMemoryStreamManager));
+    }
+
+    /// <inheritdoc />
+    public async Task UpdateInstanceBlobAsync(InstanceFileIdentifier instanceFileIdentifier, DicomDataset datasetToUpdate, CancellationToken cancellationToken)
+    {
+        EnsureArg.IsNotNull(datasetToUpdate, nameof(datasetToUpdate));
+        EnsureArg.IsNotNull(instanceFileIdentifier, nameof(instanceFileIdentifier));
+        EnsureArg.IsTrue(instanceFileIdentifier.NewVersion.HasValue, nameof(instanceFileIdentifier.NewVersion.HasValue));
+
+        Task updateInstanceFileTask = UpdateInstanceFileAsync(instanceFileIdentifier, datasetToUpdate, cancellationToken);
+        Task updateInstanceMetadataTask = UpdateInstanceMetadataAsync(instanceFileIdentifier, datasetToUpdate, cancellationToken);
+        await Task.WhenAll(updateInstanceFileTask, updateInstanceMetadataTask);
+    }
+
+    /// <inheritdoc />
+    public async Task DeleteInstanceBlobAsync(long fileIdentifier, CancellationToken cancellationToken = default)
+    {
+        _logger.LogInformation("Begin deleting instance blob {FileIdentifier}.", fileIdentifier);
+
+        Task fileTask = _fileStore.DeleteFileIfExistsAsync(fileIdentifier, cancellationToken);
+        Task metadataTask = _metadataStore.DeleteInstanceMetadataIfExistsAsync(fileIdentifier, cancellationToken);
+
+        await Task.WhenAll(fileTask, metadataTask);
+
+        _logger.LogInformation("Deleting instance blob {FileIdentifier} completed successfully.", fileIdentifier);
+    }
+
+    private async Task UpdateInstanceFileAsync(InstanceFileIdentifier instanceFileIdentifier, DicomDataset datasetToUpdate, CancellationToken cancellationToken)
+    {
+        long originFileIdentifier = instanceFileIdentifier.Version;
+        long newFileIdentifier = instanceFileIdentifier.NewVersion.Value;
+
+        KeyValuePair<string, long> block = default;
+
+        // If the file is already updated, then we can copy the file and just update the first block
+        if (instanceFileIdentifier.OriginalVersion.HasValue)
```

Comment:
I am little confused on this if and else logic. 

Do I read it like

if (previouslyUpdated)
{
 createAnotherCLone()
}
else
{
 fixTheBlocksInPlace()
}

Are'nt you suppose to copy file for the first time of update and fix the blocks in the copied version of the file?

Reply:
Yes, I am doing by calling the method StoreFileInBlocksAsync

Reply:
Sorry I am still confused. Can we sync on this tomorrow Karthik?

Reply:
I was expected

If (!OriginialVersion.HasValue)
{
  Copy()
  UpdateBlocksInTheCopy()
}

UpdatePatientModuleDataSet()